lib/stats/stats-cluster.c

```diff
@@ -78,11 +78,69 @@ stats_cluster_deinit(void)
   stats_types = NULL;
 }
 
+gboolean
+stats_cluster_key_labels_equal(StatsClusterLabel *l1, gsize l1_len, StatsClusterLabel *l2, gsize l2_len)
+{
+  if (l1_len != l2_len)
+    return FALSE;
+
+  for (gsize i = 0; i < l1_len; ++i)
+    {
+      if (strcmp(l1[i].name, l2[i].name) != 0 || strcmp(l1[i].value, l2[i].value) != 0)
+        return FALSE;
+    }
+
+  return TRUE;
+}
+
+guint
+stats_cluster_key_labels_hash(StatsClusterLabel *labels, gsize labels_len)
+{
+  guint hash = 0;
+  for (gsize i = 0; i < labels_len; ++i)
+    {
+      StatsClusterLabel *label = &labels[i];
+      hash += g_str_hash(label->name) + g_str_hash(label->value);
+    }
+
+  return hash;
+}
+
+StatsClusterLabel *
+stats_cluster_key_labels_clone(StatsClusterLabel *labels, gsize labels_len)
+{
+  StatsClusterLabel *cloned = g_new(StatsClusterLabel, labels_len);
+
+  for (gsize i = 0; i < labels_len; ++i)
+    {
+      StatsClusterLabel *label = &labels[i];
+      g_assert(label->name && label->value);
+
+      cloned[i] = stats_cluster_label(g_strdup(label->name), g_strdup(label->value));
+    }
+
+  return cloned;
+}
+
+void
+stats_cluster_key_labels_free(StatsClusterLabel *labels, gsize labels_len)
+{
+  for (gsize i = 0; i < labels_len; ++i)
+    {
+      g_free((gchar *) labels[i].name);
+      g_free((gchar *) labels[i].value);
+    }
+
+  g_free(labels);
+}
+
 StatsClusterKey *
 stats_cluster_key_clone(StatsClusterKey *dst, const StatsClusterKey *src)
 {
-  dst->legacy.component = src->legacy.component;
   dst->id = g_strdup(src->id ? : "");
+  dst->labels = stats_cluster_key_labels_clone(src->labels, src->labels_len);
+  dst->labels_len = src->labels_len;
+  dst->legacy.component = src->legacy.component;
   dst->legacy.instance = g_strdup(src->legacy.instance ? : "");
```

Comment:
This `strdup()` bothers me. It does not produce the same value after clone (`NULL` vs `""`).

Previously this was not an issue, because in `stats_cluster_single_key_set()`, if the user has given `NULL` it was replaced to `""`, which means that `src->instance` and `src->id` was never `NULL`, making this check unnecessary. 

However, now there are cases, where `legacy.instance` and later `legacy.id` are `NULL`, so this `strdup()` produces a different result, which can mess up the `equal()` function, or at least this has to be accounted for in the `equal()` function.

Reply:
I think the current implementation works. There are 2 scenarios:
 1. legacy alias was not set
     * the original key has `NULL` as `legacy.instance`
     * the cloned key gets `""` as `legacy.instance`
     * both `legacy.set` is 0
     * because of the previous point the `hash()` and `equal()` functions do not compare the legacy values, so it works
 2. legacy alias was set
     * original key has a non-NULL `legacy.instance`, because `_legacy_set()` forces it to be non-NULL.
     * the cloned key will copy the `legacy.instance`, everything works as expected

In other words the legacy values are only compared if `legacy.set` is set to 1, and when it is set to 1 (`_legacy_set()`) the problematic values are always non-NULL.